### PR TITLE
training with coodbook loss

### PR DESCRIPTION
**Pull request by glynpu**
TODO:
Ner future:
1. dataloader and augment
2. quantizer training
3. generating manifest with codebook indices

Further experiments:
1. different number of code books
2. different layer of memory embeddings
3. different weight of codebook loss
4. frame-rate mismatch between teach model and student model
5. use wav2vec as a teach model

### CODE REVIEW

**Comment by glynpu** [comment]
@zhu-han Current results show that training with code-book could converge faster and better. 
Here is the result training with librispeech-clean-100h dataset.

epoch | baseline wer | + code book loss
-- | -- | --
1 | 73.38 | 94.24
2 | 57.24 | 79.2
3 | 45.32 | 66.85
4 | 36.24 | 57.71
5 | 29.66 | 50.97
6 | 24.51 | 43.71
7 | 21.11 | 37.85
8 | 18.89 | 31.45
9 | 17.74 | 27.65
10 | 16.04 | 24.65
11 | 15.55 | 22.33
12 | 14.64 | 20.23
13 | 14.28 | 18.52
14 | 13.42 | 17.86
15 | 13.28 | 16.48
16 | 12.87 | 15.92
17 | 12.83 | 15.6
18 | 12.12 | 14.97
19 | 12.39 | 14.85
20 | 12.1 | 14.13
21 | 11.74 | 13.9
22 | 11.49 | 14.04
23 | 11.3 | 13.62
24 | 11.24 | 13.54
25 | 10.86 | 13.44
26 | 10.84 | 13.11
27 | 10.84 | 12.88
28 | 10.98 | 12.74
29 | 10.69 | 12.6
30 | 10.66 | 12.77
31 | 10.53 | 12.65
32 | 10.27 | 12.46
33 | 10.28 | 12.15
34 | 10.2 | 12.19
35 | 10.02 | 12.27
36 | 10.05 | 12.1
37 | 9.98 | 12.2
38 | 9.78 | 11.82
39 | 2.9 | 11.91
40 | 9.87 | 11.91
41 | 9.72 | 11.73
42 | 9.79 | 11.87
43 | 9.75 | 11.15
44 | 9.57 | 11.76
45 | 9.34 | 11.47
46 | 9.14 | 11.21
47 | 9.16 | 10.95
48 | 9.11 | 11.01
49 | 8.89 | 11.3
50 | 8.8 | 11.1
51 | 8.85 | 11.19
52 | 8.85 | 11.2
53 | 8.65 | 10.71
54 | 8.65 | 10.78
55 | 8.7 | 10.86
56 | 8.72 | 10.76
57 | 8.63 | 10.7
58 | 8.62 | 10.49
59 | 8.67 | 10.54
60 | 8.83 | 10.78
61 | 8.62 | 10.17
62 | 8.71 | 10.38
63 | 8.58 | 10.17
64 | 8.67 | 10.35
65 | 8.8 | 10.11
66 | 8.76 | 10.04
67 | 8.7 | 10.07
68 | 8.69 | 10.03
69 | 8.64 | 10.21
70 | 8.75 | 9.57
71 | 8.59 | 9.75
72 | 8.71 | 9.56
73 | 8.62 | 9.84
74 | 8.86 | 9.68
75 | 8.74 | 10.0
76 | 8.79 | 9.54
77 | 8.84 | 9.46

**Comment by glynpu** [comment]
Now three directions have been tried:
direction|teacher model|student mode|training data|
--|--|--|--
1|icefall released model trained with 960h| icefall model| clean-100h|
2| icefall released model trained with 960h| icefall model | full libri 960h|
3| wav2vec model | icefall model|clean-100h|

All following results are on test-clean with ctc-decoding.

Conclusions of each direction:
Results of direction 1:
The teacher model help training coverging faster and finally got a much lower wer(around 9.18% vs. 5.9%).
Results of direction 2:
The teacher model help training converging faster. But finally, the results is not significant better than baseline.
Results of direction 3:
Failed to obtain good results, though wav2vec2 model got a much lower wer than icefall released model with only ctc-decoding.

 model|wav2vec2|icefall released mode
--|--|--
wer|1.86%|2.93%

To reproduce the 1.86% result, Just run command:
```
python conformer_ctc/wav2vec_decode.py 
```

Link of wav2vec model used in this exp: https://huggingface.co/facebook/wav2vec2-large-960h-lv60-self
Link of icefall released model used in this exp: https://huggingface.co/csukuangfj/icefall-asr-librispeech-conformer-ctc-jit-bpe-500-2021-11-09

Details results of direction 1:
The teacher model help to converge faster and finally got a much lower wer(around 9.18% vs. 5.9%).

epoch | baseline 1.e | our trained model 2.c.i | our trained model 2.b.v
-- | -- | -- | --
memory | no | last | last
predictor | no codebook | (old) joint predictor | power ful
weigth | 0.0 | 9.3 | 0.3
max-duration | 100 | 100 | 100
gpus | 3 | 3 | 3
num codebooks | 4 | 4 | 16
status | (done) | done | done current best
0 | 90.04 | 100.0 | 99.5
1 | 64.72 | 64.18 | 63.6
2 | 47.81 | 41.91 | 44.12
3 | 36.94 | 28.10 | 29.17
4 | 28.29 | 21.53 | 20.83
5 | 25.28 | 17.54 | 16.86
6 | 21.3 | 15.14 | 14.42
7 | 19.42 | 14.01 | 13.38
8 | 18.35 | 13.43 | 12.65
9 | 17.37 | 12.94 | 12.12
10 | 16.41 | 12.07 | 11.17
11 | 16.07 | 11.71 | 10.85
12 | 15.39 | 11.2 | 10.5
13 | 15.15 | 11.09 | 10.1
14 | 14.68 | 10.66 | 9.84
15 | 14.37 | 10.54 | 9.4
16 | 14.67 | 10.7 | 9.58
17 | 13.82 | 10.09 | 9.18
18 | 13.68 | 9.7 | 8.71
19 | 13.39 | 9.62 | 8.62
20 | 13.14 | 9.39 | 8.58
21 | 12.96| 9.11| 8.14
22 | 12.4 | 8.87 | 8.15
23 | 12.08 | 8.68 | 7.86
24 | 11.72 | 8.45 | 7.64
25 | 11.39 | 8.22 | 7.47
26 | 11.45 | 8.12 | 7.29
27 | 11.2 | 7.88 | 7.29
28 | 11.05 | 7.84 | 6.93
29 | 10.82 | 7.86 | 6.92
30 | 10.94 | 7.63 | 6.81
31 | 10.92 | 7.52 | 6.93
32 | 10.97 | 7.6 | 6.71
33 | 10.51 | 7.59 | 6.66
34 | 10.74 | 7.52 | 6.78
35 | 10.32 | 7.45 | 6.57
36 | 10.36 | 7.5 | 6.7
37 | 10.32 | 7.39 | 6.56
38 | 10.1 | 7.29 | 6.61
39 | 10.04 | 7.25 | 6.47
40 | 10.19 | 7.29 | 6.44
41 | 10.01 | 7.23 | 6.36
42 | 10.0 | 7.24 | 6.47
43 | 9.85 | 7.08 | 6.38
44 | 10.01 | 7.06 | 6.43
45 | 9.89 | 7.06 | 6.4
46 | 9.98 | 6.98 | 6.29
47 | 9.75 | 6.96 | 6.26
48 | 9.79 | 7.06 | 6.42
49 | 9.71 | 6.99 | 6.24
50 | 9.8 | 7.09 | 6.18
51 | 9.77 | 7.01 | 6.08
52 | 9.69 | 7.11 | 5.96
53 | 9.59 | 7.07 | 6.04
54 | 9.47 | 7.01 | 6.18
55 | 9.64 | 7.0 | 6.08
56 | 9.68 | 6.95 | 6.07
57 | 9.68 | 6.94 | 6.08
58 | 9.72 | 6.89 | 6.03
59 | 9.38 | 6.89 | 5.99
60 | 9.55 | 6.9 | 6.02
61 | 9.61 | 6.9 | 5.97
62 | 9.43 | 6.88 | 6.0
63 | 9.55 | 6.88 | 5.97
64 | 9.4 | 6.93 | 6.06
65 | 9.54 | 7.0 | 5.94
66 | 9.49 | 6.81 | 6.06
67 | 9.43 | 6.86 | 6.07
68 | 9.33 | 6.89 | 5.97
69 | 9.14 | 6.79 | 5.87
70 | 9.21 | 6.73 | 5.91
71 | 9.07 | 6.8 | 5.95
72 | 9.17 | 6.88 | 5.8
73 | 9.34 | 6.92 | 5.93
74 | 9.18 | 6.96 | 5.83
75 | 9.41 | 6.87 | 5.79
76 | 9.21 | 6.79 | 5.97
77 | 9.18 | 6.77 | 5.83




Detail Results of direction 2:
The teacher model help training converges faster. But finally, the results is not significant better than baseline.

epoch | full libri baseline | full libri  2.e.ii
-- | -- | --
memory | no | last
predictor | no codebook | powerful
weigth | 0.0 | 0.3
max-duration | 200 | 270
gpus | 4 | 3
valid batch duration | 800=4*200 | 810=3*270
batches per epoch | 13500 | 13200
num codebooks | 0 | 16
status | (done) 2.5 hours/ epoch | training 4 hours/ epoch
0 | 31.96 | 26.89
1 | 13.09 | 9.97
2 | 9.41 | 7.67
3 | 8.29 | 6.89
4 | 7.67 | 6.47
5 | 7.0 | 6.08
6 | 5.75 | 5.50
7 | 5.45 | 5.12
8 | 5.13 | 4.76
9 | 4.81 | 4.49
10 | 4.68 | 4.37
11 | 4.46 | 4.35
12 | 4.38 |  4.17
13 | 4.4 |  4.12
